parquet-column/src/main/java/org/apache/parquet/column/values/bloom/Bloom.java

```diff
@@ -0,0 +1,485 @@
+   */
+  public BytesInput getBytes() {
+    inputs.clear();
+
+    // Add number of bytes.
+    inputs.add(BytesInput.fromInt(numBytes));
+
+    // Add hash strategy
+    inputs.add(BytesInput.fromInt(this.bloomFilterHash.ordinal()));
+
+    // Add bloom filter algorithm
+    inputs.add(BytesInput.fromInt(this.bloomFilterAlgorithm.ordinal()));
+
+    // Add bitset.
+    inputs.add(BytesInput.from(bitset, 0, bitset.length));
+
+    return BytesInput.concat(inputs);
+  }
+
+  private int[] setMask(int key) {
+    // The block based algorithm needs 8 odd SALT values to calculate index
+    // of bit to set in block.
+    final int SALT[] = {0x47b6137b, 0x44974d91, 0x8824ad5b, 0xa2b7289d,
+      0x705495c7, 0x2df1424b, 0x9efc4947, 0x5c6bfb31};
+
+    int mask[] = new int[8];
+
+    for (int i = 0; i < 8; ++i) {
+      mask[i] = key * SALT[i];
+    }
+
+    for (int i = 0; i < 8; ++i) {
+      mask[i] = mask[i] >> 27;
+    }
+
+    for (int i = 0; i < 8; ++i) {
+      mask[i] = 0x1 << mask[i];
+    }
+
+    return mask;
+  }
+
+  /**
+   * Add an element to bloom filter, the element content is represented by
+   * the hash value of its plain encoding result.
+   * @param hash hash result of element.
+   */
+  public void bloomInsert(long hash) {
+    int bucketIndex = (int)(hash >> 32) & (numBytes / BYTES_PER_BUCKET - 1);
+    int key = (int)hash;
+
+    // Calculate bit mask for bucket.
+    int mask[] = setMask(key);
+
+    for (int i = 0; i < 8; i++) {
+      int bitsetIndex = bucketIndex * BYTES_PER_BUCKET + i * 4;
+      bitset[bitsetIndex] |= (byte)(mask[i] >>> 24);
+      bitset[bitsetIndex + 1] |= (byte)(mask[i] >>> 16);
+      bitset[bitsetIndex + 2] |= (byte)(mask[i] >>> 8);
+      bitset[bitsetIndex + 3] |= (byte)(mask[i]);
+    }
+  }
+
+  /**
+   * Determine where an element is in set or not.
+   * @param hash the hash value of element plain encoding result.
+   * @return false if element is must not in set, true if element probably in set.
+   */
+  public boolean bloomFind(long hash) {
+    if (!elements.isEmpty()) {
+      flush();
+    }
+
+    int bucketIndex = (int)(hash >> 32) & (numBytes / BYTES_PER_BUCKET - 1);
+    int key = (int)hash;
+
+    // Calculate bit mask for bucket.
+    int mask[] = setMask(key);
+
+    for (int i = 0; i < 8; i++) {
+      byte set = 0;
+      int bitsetIndex = bucketIndex * BYTES_PER_BUCKET + i * 4;
+      set |= bitset[bitsetIndex] & ((byte)(mask[i] >>> 24));
+      set |= bitset[bitsetIndex + 1] & ((byte)(mask[i] >>> 16));
+      set |= bitset[bitsetIndex + 2] & ((byte)(mask[i] >>> 8));
+      set |= bitset[bitsetIndex + 3] & ((byte)mask[i]);
+      if (0 == set) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  /**
+   * Bloom filter bitset can be created lazily, flush() will set bits for
+   * all elements in cache. If bitset was already created and set, it do nothing.
+   */
+  public void flush() {
+    if (!elements.isEmpty() && bitset == null) {
```

Comment:
Why does this silently do nothing if !elements.isEmpty() but bitset != null? Should it throw an exception?

Reply:
The original bloomfilter value writer has two path, one is writing values to a hash set as cache, and finalize bloom filter when finalize the column chunk. Another is setting bits immediately when writing values which does not need a flush since the bitset is ready.

Reply:
I don't know what the "original bloom filter value writer" is. This is the first one I have seen.

Please add a check here for this case, or, preferably, move the buffering BF construction out of the BF class.